### PR TITLE
closes 5111 serve webp for youtube

### LINKED ISSUE #5111: Serve YouTube Thumbnail in WebP format.
**Is your feature request related to a problem? Please describe.**
YouTube can serve thumbnails in WebP format, but WP Rocket Lazyload is serving only JPG.

**Describe the solution you'd like**
Making a few changes in the URL, a WebP image can be served.

Example:

JPG: https://i.ytimg.com/vi/Fgvokdu74fw/hqdefault.jpg
WebP: https://i.ytimg.com/vi_webp/Fgvokdu74fw/hqdefault.webp

The only differences are: "vi_webp" and ".webp"

After some test with several videos, all of them had the WebP image available.

**Describe alternatives you've considered**

**Additional context**
Doing this, we can fix this kind of issues:
image

Also, in some images there are big differences in file size:
image

### PR DESCRIPTION
**Pull request by CrochetFeve0251**
## Description

Solve an issue where the extension from the youtube thumbnail was still jpg even when we activated webp.

For that we changed the lazyload module and the webp subscriber to make the extension dynamic.

Fixes #5111

## Type of change

Please delete options that are not relevant.

- [ ] Enhancement (non-breaking change which improves an existing functionality)

## Is the solution different from the one proposed during the grooming?

No.

## How Has This Been Tested?

- [ ] Automated Test

# Checklist:

Please delete the options that are not relevant.

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by jeawhanlee — APPROVED** [review]

**Review by remyperona — COMMENTED** [review]
Isn't there change missing in the RocketLazyload\Assets class to use the custom extension value?

**Comment by CrochetFeve0251** [comment]
> Isn't there change missing in the RocketLazyload\Assets class to use the custom extension value?

I made the change on the lazyload repository directly: https://github.com/wp-media/rocket-lazyload-common/pull/109.
Should I modify here too?

**Comment by remyperona** [comment]
Ok, next step is to create a new release there, so that it can be pulled on this repo

**Comment by CrochetFeve0251** [comment]
> Ok, next step is to create a new release there, so that it can be pulled on this repo

Ok I done it

**Comment by remyperona** [comment]
You can pull the updated package with composer update and commit it to finalize the PR on that side

**Comment by CrochetFeve0251** [comment]
> You can pull the updated package with composer update and commit it to finalize the PR on that side

Ok I done it

**Review by remyperona — APPROVED** [review]

**Comment by vmanthos** [comment]
@CrochetFeve0251 

1. Currently, **the WebP thumbnail URL is broken**. 
Instead of being:
`https://i.ytimg.com/vi_webp/PEp711KyvgI/hqdefault.webp`

it's: `https://i.ytimg.com/vi/PEp711KyvgI/hqdefault.webp`.

Note that after `https://i.ytimg.com/` there is `vi`. It should be `vi_webp` when the `webp` thumbnail is to be displayed.

2. We should **safeguard** `rocket_lazyload_youtube_thumbnail_extension` **against invalid values**, e.g. using `null` results in a PHP fatal error. If anything invalid is returned we can **use** `jpg` **as a fallback**.

Can you please look into these? :pray:

File: inc/Engine/Media/Lazyload/Subscriber.php
Comment (vmanthos) [inline inc/Engine/Media/Lazyload/Subscriber.php:1]: I believe this should be:
```PHP
if ( ! is_string( $extension ) ) {

**Comment by remyperona** [comment]
There is a branch sync issue on this PR currently

**Review by remyperona — APPROVED** [review]

**Comment by vmanthos** [comment]
@CrochetFeve0251 While doing the final checks I found that if:
1. The WebP add-on is enabled `&&`
2. The following is added to the filter:
```PHP
add_filter( 'rocket_lazyload_youtube_thumbnail_extension', function( $extension ){
	return null;
});
```
an `error 500` takes place, and the following is added to the `debug.log`:
```PHP
PHP Fatal error:  Uncaught TypeError: WP_Rocket\Addon\WebP\Subscriber::change_youtube_thumbnail(): Argument #1 ($extension) must be of type string, null given, called in /shared/httpd/wprocketest/htdocs/wp-includes/class-wp-hook.php on line 308 and defined in /shared/httpd/wprocketest/htdocs/wp-content/plugins/wp-rocket/inc/Addon/WebP/Subscriber.php:592
Stack trace:
#0 /shared/httpd/wprocketest/htdocs/wp-includes/class-wp-hook.php(308): WP_Rocket\Addon\WebP\Subscriber->change_youtube_thumbnail(NULL)
#1 /shared/httpd/wprocketest/htdocs/wp-includes/plugin.php(205): WP_Hook->apply_filters(NULL, Array)
#2 /shared/httpd/wprocketest/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Media/Lazyload/Subscriber.php(246): apply_filters('rocket_lazyload...', 'jpg')
#3 /shared/httpd/wprocketest/htdocs/wp-includes/class-wp-hook.php(308): WP_Rocket\Engine\Media\Lazyload\Subscriber->insert_youtube_thumbnail_script('')
#4 /shared/httpd/wprocketest/htdocs/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters(NULL, Array)
#5 /shared/httpd/wprocketest/htdocs/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#6 /shared/httpd/wprocketest/htdocs/wp-includes/general-template.php(3065): do_action('wp_footer')
#7 /shared/httpd/wprocketest/htdocs/wp-content/themes/twentytwentyone/footer.php(78): wp_footer()
#8 /shared/httpd/wprocketest/htdocs/wp-includes/template.php(783): require_once('/shared/httpd/w...')
#9 /shared/httpd/wprocketest/htdocs/wp-includes/template.php(718): load_template('/shared/httpd/w...', true, Array)
#10 /shared/httpd/wprocketest/htdocs/wp-includes/general-template.php(92): locate_template(Array, true, true, Array)
#11 /shared/httpd/wprocketest/htdocs/wp-content/themes/twentytwentyone/page.php(25): get_footer()
#12 /shared/httpd/wprocketest/htdocs/wp-includes/template-loader.php(106): include('/shared/httpd/w...')
#13 /shared/httpd/wprocketest/htdocs/wp-blog-header.php(19): require_once('/shared/httpd/w...')
#14 /shared/httpd/wprocketest/htdocs/index.php(17): require('/shared/httpd/w...')
#15 {main}
  thrown in /shared/httpd/wprocketest/htdocs/wp-content/plugins/wp-rocket/inc/Addon/WebP/Subscriber.php on line 592
```

Is there a way to prevent this?

**Review by vmanthos — APPROVED** [review]
This is working as expected! :+1: 

TestRail report: 
[testrail-report-478.pdf](https://github.com/wp-media/wp-rocket/files/11330591/testrail-report-478.pdf)

**NOTE**

Using code like the following:
```PHP
add_filter( 'rocket_lazyload_youtube_thumbnail_extension', function( $extension ){
	return jpg;
});
```
will throw a PHP fatal error:
```PHP
PHP Fatal error:  Uncaught Error: Undefined constant "jpg" in /shared/httpd/wprocketest/htdocs/wp-content/plugins/code-snippets/php/snippet-ops.php(504) : eval()'d code:2
Stack trace:
#0 /shared/httpd/wprocketest/htdocs/wp-includes/class-wp-hook.php(308): {closure}('jpg')
#1 /shared/httpd/wprocketest/htdocs/wp-includes/plugin.php(205): WP_Hook->apply_filters('jpg', Array)
#2 /shared/httpd/wprocketest/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Media/Lazyload/Subscriber.php(246): apply_filters('rocket_lazyload...', 'jpg')
#3 /shared/httpd/wprocketest/htdocs/wp-includes/class-wp-hook.php(308): WP_Rocket\Engine\Media\Lazyload\Subscriber->insert_youtube_thumbnail_script('')
#4 /shared/httpd/wprocketest/htdocs/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters(NULL, Array)
#5 /shared/httpd/wprocketest/htdocs/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#6 /shared/httpd/wprocketest/htdocs/wp-includes/general-template.php(3065): do_action('wp_footer')
#7 /shared/httpd/wprocketest/htdocs/wp-content/themes/twentytwentyone/footer.php(78): wp_footer()
#8 /shared/httpd/wprocketest/htdocs/wp-includes/template.php(783): require_once('/shared/httpd/w...')
#9 /shared/httpd/wprocketest/htdocs/wp-includes/template.php(718): load_template('/shared/httpd/w...', true, Array)
#10 /shared/httpd/wprocketest/htdocs/wp-includes/general-template.php(92): locate_template(Array, true, true, Array)
#11 /shared/httpd/wprocketest/htdocs/wp-content/themes/twentytwentyone/single.php(50): get_footer()
#12 /shared/httpd/wprocketest/htdocs/wp-includes/template-loader.php(106): include('/shared/httpd/w...')
#13 /shared/httpd/wprocketest/htdocs/wp-blog-header.php(19): require_once('/shared/httpd/w...')
#14 /shared/httpd/wprocketest/htdocs/index.php(17): require('/shared/httpd/w...')
#15 {main}
  thrown in /shared/httpd/wprocketest/htdocs/wp-content/plugins/code-snippets/php/snippet-ops.php(504) : eval()'d code on line 2
```

After discussing this with @CrochetFeve0251, this is a user's mistake and there isn't a way to prevent this from happening.

**Pull request merged**